src/genetic.c

```diff
@@ -0,0 +1,666 @@
+ *
+ *
+ * @param population this array stores all the values of the population
+ * @param newGeneration this array stores all of the created children
+ * @param bestFitValues this array stores the parameters of the current best
+ * solution
+ * @param secondBestValues this array stores the parameters of the current
+ * second best solution
+ * @param arraySize this indicates the size of the population
+ * @param dimensions the number of parameters in the function to optimize
+ */
+static void
+replacePopulation(genetic_int *population, genetic_int *newGeneration,
+                  genetic_int *bestFitValues, genetic_int *secondBestValues,
+                  const unsigned int arraySize, const unsigned int dimensions) {
+
+  const unsigned int nextGenerationStartingIndex = 2 * dimensions;
+
+  // We copy the elite values before copying the rest of the values from the
+  // next generation
+
+  memcpy(population, bestFitValues, dimensions * sizeof(genetic_int));
+  memcpy(population + dimensions, secondBestValues,
+         dimensions * sizeof(genetic_int));
+  memcpy(population + nextGenerationStartingIndex, newGeneration,
+         arraySize * sizeof(genetic_int));
+}
+
+/**
+ * This function runs a genetic algorithm to minimize a function by finding the
+ * best parameters possible the epsilon is the goal fitness and the execution
+ * stops once that it is attained the other cutoff is the maximumIterationCount
+ * parameter
+ *
+ * @param bestFitValues this array stores the parameters of the best solution
+ * @param parameterCount this function indicates the number of parameters in the
+ * efunction to optimize
+ * @param epsilon the target fitness to achieve
+ * @param mutationChance the chance that a created child will have a mutation
+ * @param generationSize the amount of solutions to add into the population
+ * @param maximumIterationCount the maximum amount of generations that will be
+ * created
+ * @param tourneySize the number of solutions that are chosen in the tourney ,
+ * must be smallere than the population size
+ * @param evaluationFunction the function that is used to
+ * @param lowMemoryMode if set to 1 the algorithm will be slower and call the
+ * evaluation function more often
+ * @return real the fitness of the best solution
+ */
+genetic_real geneticAlgorithm(genetic_real *bestFitValues,
+                              const unsigned int parameterCount,
+                              const genetic_real epsilon,
+                              const genetic_real mutationChance,
+                              unsigned int generationSize,
+                              const unsigned int maximumIterationCount,
+                              const unsigned int tourneySize,
+                              fitness_evaluation_function evaluationFunction,
+                              const unsigned int lowMemoryMode) {
+
+  if (generationSize % 2)
+    generationSize += 1;
+
+  if (tourneySize > generationSize)
+    exit(1);
```

Comment:
You should not exit the program if you encounter an error. You should return an error code to the caller